### ISSUE TITLE
Idle CPU usage anomaly

### ISSUE THREAD

**Issue by bolle732** [stsp/dosemu2#211]
On Ubuntu 16.04 LTS 32-bit. When running PC/GEOS with DOSEmu 1.4 and it is doing "nothing" (idling), the 'dosemu.bin' process uses about 3% CPU. Running the same PC/GEOS under DOSEmu2 with 'vm86' or 'kvm', the 'dosemu.bin' process is using over 60% of a CPU core.

I've tried to use FDAPM and APM drivers inside PC/GEOS, but DOSEmu doesn't seems to emulate this function.

The high CPU usage can be observed too on a 64-bit installation with 'kvm'. Even running the demo of Commander Keen on a new notebook uses more the 100% of a CPU core.

What could make that big difference between the two versions ?

Regarding power, it would be really nice to get back to the lower CPU usage when the DOS program is not demanding CPU power. Especially on a notebook, the battery time is suffering with this behaviour.

**Comment by stsp** [comment]
It would really help if you bisect that.
Thanks!

**Comment by bolle732** [comment]
I'll try it next week. My current old laptop at home is to slow for that.

**Comment by jschwartzenberg** [comment]
I observe high CPU usage in general as well with dosemu2. I should check against 1.4.

**Comment by bolle732** [comment]
Having loaded only FreeDOS, DOSEmu2 stay at ~5%. This is ok. I see the high CPU usage especially if a program switched to graphic mode.

**Comment by jschwartzenberg** [comment]
Yes this is exactly the same I'm seeing as well.

**Comment by jschwartzenberg** [comment]
It's not that easy for me to reproduce this. I see at least that kvm seems to have very little effect on reducing the CPU usage. I think it would be a good idea to run different performance benchmarks in 1.4 and 2. Not sure if this would catch anything regarding idling though...

**Comment by bolle732** [comment]
Sorry, but I hadn't currently the time to do the bisect.

Regarding KVM, it would be difficult, as I think that KVM started to work after the relevant commits. Have to do it on vm86. Hopefully next week.

**Comment by bolle732** [comment]
So, after fighting with problems during bisect with different versions and kernels, I decided to give 'strace' and 'ltrace' a chance.

I think there is an issue with pthread mutex and semaphore wait. After the run of 'ltrace' on DOSEmu2, PC/GEOS has freezed. This is possibly due to 'ltrace' demanding lot of CPU too and the CPU usage of DOSEmu2 lowered from >50% to about ~30%. The freeze happens with 'vm86' and 'kvm'.

============================================================
DOSEmu1 with VM86:
Top reports 2 - 3% CPU PC/GEOS is started up and idling.
```
# timeout 30 strace -c -p 1438
strace: Process 1438 attached
strace: Process 1438 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 75.84    0.072000          30      2439      2439 rt_sigsuspend
 11.54    0.010951           0     56595           mprotect
  4.00    0.003799           0     10496         8 _newselect
  3.39    0.003216           0     16706           vm86
  1.89    0.001791           0      5293           mmap2
  1.77    0.001680           0      5455      2447 sigreturn
  1.40    0.001330           0      3609      3443 recvmsg
  0.11    0.000101           1       166           poll
  0.07    0.000065           0       166           writev
------ ----------- ----------- --------- --------- ----------------
100.00    0.094933                100925      8337 total

# timeout 30 ltrace -c -p 1438
% time     seconds  usecs/call     calls      function
------ ----------- ----------- --------- --------------------
 27.59    6.843729         284     24088 mprotect
 19.04    4.722637         213     22078 gettimeofday
 18.96    4.703755         571      8236 syscall
 18.61    4.615300         417     11052 select
 13.95    3.460363         207     16702 strcmp
  1.28    0.318122         306      1038 mmap64
  0.38    0.094395        3933        24 sigsuspend
  0.09    0.023021         201       114 sysconf
  0.08    0.019749         274        72 __errno_location
  0.02    0.004959         206        24 sigemptyset
------ ----------- ----------- --------- --------------------
100.00   24.806030                 83428 total
```
============================================================
DOSEmu2 with VM86:
Top reports 52 - 54% CPU PC/GEOS is started up and idling.
```
# timeout 30 strace -c -p 1981
strace: Process 1981 attached
strace: Process 1981 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 90.24    0.673612         254      2655      2655 rt_sigsuspend
  6.47    0.048288          11      4431        12 futex
  1.73    0.012931           0     48672           mprotect
  1.36    0.010178           1     17003           vm86
  0.07    0.000554           0      4077           mremap
  0.07    0.000538           0      6791      2660 rt_sigreturn
  0.03    0.000254           0      5570           _newselect
  0.02    0.000142           0      2299           clock_gettime
  0.00    0.000000           0         7           writev
  0.00    0.000000           0         7           poll
  0.00    0.000000           0      3276           gettid
  0.00    0.000000           0         1           sendmsg
  0.00    0.000000           0      2306      2300 recvmsg
------ ----------- ----------- --------- --------- ----------------
100.00    0.746497                 97095      7627 total

# timeout 30 ltrace -c -p 1981
% time     seconds  usecs/call     calls      function
------ ----------- ----------- --------- --------------------
 42.12   32.250206       15739      2049 sem_wait
 15.50   11.872378        1279      9279 pthread_mutex_lock
 13.71   10.494676        1131      9277 pthread_mutex_unlock
  8.71    6.666689        6510      1024 usleep
  7.10    5.435149        1657      3279 syscall
  6.32    4.837384        2359      2050 sem_post
  2.69    2.062030        1006      2049 pthread_setcancelstate
  1.47    1.123124        1078      1041 select
  1.16    0.886685         865      1025 memcpy
  1.13    0.866366         845      1025 _setjmp
  0.05    0.035804         942        38 strcmp
  0.03    0.022857        1142        20 clock_gettime
  0.02    0.012344         726        17 __errno_location
  0.01    0.006169         881         7 feenableexcept
------ ----------- ----------- --------- --------------------
100.00   76.571861                 32180 total
```
============================================================
DOSEmu2 with KVM:
Top reports 52 - 54% CPU PC/GEOS is started up and idling.
```
# timeout 30 strace -c -p 1601
strace: Process 1601 attached
strace: Process 1601 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 95.68    0.865384         293      2952      2952 rt_sigsuspend
  2.22    0.020109           5      4331        11 futex
  1.26    0.011407           1     14853        35 ioctl
  0.46    0.004180           1      5627           _newselect
  0.25    0.002274           0     28742           mprotect
  0.10    0.000886           0      2255           mremap
  0.01    0.000118           0      2358           clock_gettime
  0.01    0.000077           0      3275           gettid
  0.00    0.000000           0         7           writev
  0.00    0.000000           0         7           poll
  0.00    0.000000           0      3275      2988 rt_sigreturn
  0.00    0.000000           0         1           sendmsg
  0.00    0.000000           0      2364      2358 recvmsg
------ ----------- ----------- --------- --------- ----------------
100.00    0.904435                 70047      8344 total

# timeout 30 ltrace -c -p 1601
% time     seconds  usecs/call     calls      function
------ ----------- ----------- --------- --------------------
 41.96   31.971524       15512      2061 sem_wait
 15.32   11.673350        1250      9337 pthread_mutex_lock
 13.63   10.381629        1111      9337 pthread_mutex_unlock
  8.85    6.741593        6538      1031 usleep
  6.95    5.294979        1619      3269 syscall
  6.77    5.155930        2500      2062 sem_post
  2.66    2.025254         981      2063 pthread_setcancelstate
  1.52    1.157953        1094      1058 select
  1.14    0.869229         843      1031 memcpy
  1.09    0.831804         806      1031 _setjmp
  0.05    0.035782         941        38 strcmp
  0.03    0.023158        1157        20 clock_gettime
  0.03    0.020254         750        27 __errno_location
  0.01    0.005801        1160         5 ioctl
  0.00    0.003559         711         5 feenableexcept
------ ----------- ----------- --------- --------------------
100.00   76.191799                 32375 total
```

**Comment by stsp** [comment]
It seems geos updates the screen very frequently,
regardless of whether it was really updated.
dosemu2 is very aggressive on rendering, so it
takes these updates very seriously, while dosemu1
can just ignore everything.
So I don't consider this a bug, just a room for an
improvement. We can re-check things with memcmp()
probably to exclude those fake updates...
In a mean time you can fix your program.

**Comment by andrewbird** [comment]
Interestingly, memcmp can be sse accelerated depending upon platform/glibc/gcc etc, so a useful speedup may occur.

**Comment by stsp** [comment]
Maybe we could also use kernel KSM.
It does pretty much the same thing, and the
difference with the user-space solution is that
kernel will not memcmp() the pages that are
still CoW (i.e. definitely unmodified), and can
re-CoW same pages faster than user-space.
But perhaps this is a too small bonus to mess
with KSM.

**Comment by bolle732** [comment]
Sorry for my late answer, but I can't fix PC/GEOS myself as I don't have the source to do that. If DOSEmu2 would have an API for para-virtualized devices, then I could get interested (I have access to the source code of a GDI driver for running it under Windows)...

Running PC/GEOS at 1024x768x64k, the CPU usage is about 50%. If I use 640x480x16 graphic mode, the CPU usage is about 15%. Commander Keen 4 with EGA/VGA uses about 10% CPU on the menu screen. Therefore, it shouldn't be a problem related to PC/GEOS only.

In my eyes, there would be a lot of other programs working that way, as I believe the graphic card is something a programmer used as fast as he can. Some of them have programmed a system without video RAM at all ;-) Mostly, no one had in the mind, that an emulator will be thrilled with this. And power saving wasn't a question too. Unfortunately.

DOSEmu 1.x could handle that very well. I don't complain about 10% difference, but a factor of 10 to 20 is little bit too much. Having a notebook with 2 cores and 1 core is always over 50% will reduce the run time on battery a lot. Did the old '$_X_updfreq' something in this area?

I won't stress here and you decide what will be done and what not. But my feeling is that the way how DOSEmu2 does it currently isn't the optimal one.

**Comment by stsp** [comment]
Yes, dosemu1 handled _that_ very well
(by not doing anything at all), but it did much
worse on an apps that really wanted to render
something: the frame rate was much, much lower.
Yes, it would be nice to get this improved, but
this will require some additions to memory mapping
library, and quite some work on vgaemu side.
Not something I am interested for the next pre
release, but the patches are always welcome. :)

**Comment by bolle732** [comment]
No problem that this isn't a first priority citizen. To be able to configure skipping updates of the screen would be sufficient in that case.

Is the pthread mutex used to synchronize the access to the "video" memory ?

**Comment by stsp** [comment]
Mostly not. The DOS prog can write to video
memory concurrently with dosemu reading from
it, and dosemu does never write to it, so no thread
synchronization is needed unless the video mode
is being changed.
There is a write-protection to track dirty
state - that one should be removed and
replaced with the more accurate techniques.
I am thinking of the CoW mirrors+memcmp()
in a separate thread.

**Comment by bolle732** [comment]
I can confirm the CPU usage is ways better now. On my current machine with an idle "dosemu.bin" and PC/GEOS started in it, CPU usage has dropped from around 48% to 4%.

Great work, my notebook battery will honor it!

**Comment by stsp** [comment]
Unfortunately I had to disable the fix as
it breaks other things. I think the only proper
solution is to complete rewrite of the vga update
code. It should properly track page reachability,
handle different types of mem wrap (mem.wrap
and line_compare are very different and can
co-exist), etc. I'll give it a try when I have more
time.

Btw, can you still reproduce the sluggishness in
CGA mode?

**Comment by bolle732** [comment]
I've double checked the current devel a10fb304b869391a420b5d912df4605ffd94f2a3.

Neither the mouse is sluggish under CGA nor is the CPU consumption high ?!? Now, I'm puzzled. Need to verify at home on my other system...

**Comment by stsp** [comment]
Above I pointed the commit that I think have
fixed the sluggishness problems.
But as for the CPU consumption - I'd be surprised
if its still not there. Will re-check myself as I think
I was able to reproduce it.

**Comment by bolle732** [comment]
Ok, I've rechecked CGA mode:

d3ef1fce80ed4aa7ebbc64bec0bd1aba009e7d0f -> Mouse sluggish (ERROR: callback queue overflow, X events)
33f22dd0301e3b58d646e0aeaf33aa38ac539579 -> Mouse OK

So, indeed that fixed the problem!

**Comment by bolle732** [comment]
Regarding CPU usage: Even on my second system, after around 20 seconds of idle desktop, the CPU usage drops down to 3 - 5%. But I must say, I don't have access to the system I tested in the first post.

**Comment by stsp** [comment]
OK, I did more investigations.
It appears the problem was happening only
when SDL hw rendering was used. So since
the sw rendering was defaulted, it have fixed
itself. I still can reproduce it by disabling $_SDL_swrend,
but not after commit that returns to sdl drawing
surface. After that commit, reproducing the CPU
load problem is completely impossible, no matter
what options are used and what fixes are reverted.

**Comment by stsp** [comment]
I added the debug code that would write
`ERROR: update with zero rects count`
into the -D+v log if that bug still pops up.
This is the only method now we can detect
it with, as the CPU load was fixed by many
other changes.
After fixing a few small bugs, it appears
this check doesn't trigger all too often now.
So there seem to be so many changes
recently that make this problem to go away,
that we may not even need the heavy-weight
"proper" fix after all...
